src/nginx/module.cc

```diff
@@ -49,6 +49,39 @@ namespace google {
 namespace api_manager {
 namespace nginx {
 
+namespace {
+
+// Rewrite url based on api base_path. Returns NGX_OK if rewrite was successful
+// or rewrite is not required. Otherwise, returns NGX_ERROR.
+// When this function returns NGX_ERROR, nginx should return 404 error.
+ApiBasepathRewriteAction ngx_esp_rewrite_url(ngx_http_request_t *r,
+                                             ngx_esp_loc_conf_t *lc) {
+  std::string destination_url;
+  ApiBasepathRewriteAction action =
+      lc->esp->ReWriteURL(ngx_str_to_std(r->unparsed_uri), &destination_url);
+
+  if (action == ApiBasepathRewriteAction::REWRITE) {
+    ngx_str_copy_from_std(r->pool, destination_url, &r->unparsed_uri);
+
+    ngx_str_copy_from_std(r->pool, ngx_str_to_std(r->method_name) + " " +
+                                       destination_url + " " +
+                                       ngx_str_to_std(r->http_protocol),
+                          &r->request_line);
+  } else if (action == ApiBasepathRewriteAction::REJECT) {
+    return action;
+  }
+
+  std::string uri;
+  action = lc->esp->ReWriteURL(ngx_str_to_std(r->uri), &uri);
```

Comment:
It is good idea to call ReWritURL twice for uri and unparsed_uri.    
But we need to make sure they return values are the same, 
  If different, should be the same policy as prefix not matched.  most likely prefix has url escape in it